### PR TITLE
Two more forms of hopping rates: D_s * L_{i,j} and D_{s,i} * L_{i,j}

### PR DESCRIPTION
**Pull request by Vilin97**
Also reworked little bits of `nsm.jl` and `DirectCRDirect` for compatibility, and factored out more common functions from `HopRates`

### CODE REVIEW

**Review by isaacsas — CHANGES_REQUESTED** [review]
I still need to review hop_rates.jl, but I'm going to wait till this is updated to master after the other PR is merged.

File: test/spatial/diffusion.jl
Comment (Vilin97) [reply test/spatial/diffusion.jl:64]: `rand(algs)` so that tests don't take twice longer than they do. If that's bad, I can change it to use both.
What is the benefit of `StableRngs` with a fixed seed? To avoid the very small probability that tests don't get passed because of a 1 in a million fluke?

File: src/spatial/hop_rates.jl
Comment (Vilin97) [reply src/spatial/hop_rates.jl:36]: Both exist. `update_hop_rates!` just calls `update_hop_rate!` several times.

File: src/spatial/hop_rates.jl
Comment (Vilin97) [reply src/spatial/hop_rates.jl]: Got it. I put it back now.

File: test/spatial/diffusion.jl
Comment (isaacsas) [reply test/spatial/diffusion.jl:64]: Yes, let's test all the algorithms for now. If it starts taking too long we can change it to just use a specific fixed one of them for most tests (I wouldn't want to use a random algorithm in any case).

The benefit of using a fixed seed and StableRNG is it makes the test deterministic, so if it suddenly breaks we know it is definitely a code issue. For your tests it probably doesn't matter, but we've definitely had issues with other tests occasionally breaking because of the rng hitting a degenerate case.

File: src/spatial/hop_rates.jl
Comment (isaacsas) [inline src/spatial/hop_rates.jl:339]: HopRatesGeneralGridDsiLij

**Review by isaacsas — CHANGES_REQUESTED** [review]

File: src/spatial/hop_rates.jl
Comment (Vilin97) [reply src/spatial/hop_rates.jl:208]: This is a typo. Removed.

**Pull request merged**
